packages/swingset-liveslots/src/virtualObjectManager.js

```diff
@@ -263,6 +263,20 @@ function insistDurableCapdata(vrm, what, capdata, valueFor) {
   });
 }
 
+function insistSameCapData(oldCD, newCD) {
+  if (oldCD.body !== newCD.body) {
+    Fail`durable Kind stateShape mismatch (body)`;
+  }
```

Comment:
This is assuming the encoding of the old capdata is the same as the encoding of the new cap data. Technically the 2 shapes could be the same, but "upgrading" e.g. from qclass based capdata to smallcaps.

If we're ok with this limitation, let's document it. We may have similar assumptions in other places, but I'd like us to start being cognizant of them.

A forward compatible check would instead re-encode and stringify both shapes the same way and compare the serialized strings of the shapes's cap data.

Reply:
Yeah, I can add a `serialize(deserialize(oldShapeCD))`, but it'll still get a false-mismatch on assignment-order differences like `{ x: M.string(), y: M.number() }` vs `{ y: M.number(), x: M.string() }`. To consider those identical, we'd need to replace the JSON.encode in our marshaller with the deterministic one from `djson.js` (which is what we do for kernel-side transcript entries), and that feels a bit slow and intrusive to do on the vat side.

Or, I supposed we could do `djson.encode(JSON.decode(serialize(deserialize(oldShapeCD))))` and compare it against `djson.encode(JSON.decode(serialize(newShape)))`.. doing something special in the VOM, but not affecting any other uses of marshal.

As a reviewer, would you approve such a monster?

Reply:
Actually, I'm just going to document the limitation, and not try heroics to re-serialize it. If/when we introduce a new marshalling format, we can change liveslots to do the reserialization in the same commit that begins to tolerate the new format.

Reply:
> it'll still get a false-mismatch on assignment-order differences like `{ x: M.string(), y: M.number() }` vs `{ y: M.number(), x: M.string() }`

The encoding is supposed to canonicalize this kind of things. In this case, marshal does sort the property keys lexically.